[bbryant824]iP

bbryant824 opened the pull request:
# KobeBee ~ 

> "In the words of Kobe Bryant, Mamba Out. But in the words of us, not forgotten."  
> – _Lebron James_([source](https://web.telegram.com)

Kobee~ frees your mind from having to remember things you need to do. It's designed to be:

- Text-based 📝
- _Easy to learn_ 🚀
- FAST – Super Fast! ⚡

## 🚀 Getting Started with Kobe

To get started, just follow these simple steps:

1. Download Kobe from [this link](https://web.telegram.com).
2. Double-click the downloaded file.
3. Add your tasks and let Kobe manage them for you.
4. Sit back and relax – It's FREE! 🎉

## 🔧 Features

Bong helps you manage your tasks efficiently. Here's what's currently available and what's coming soon:

- [x] Managing tasks: Add, remove, or update tasks effortlessly.
- [x] Managing deadlines: Add, remove, or update tasks effortlessly.
- [x] Reminders _(coming soon)_ ⏰

## ✨ Example Usage

Here's how you can add a task in Kobe:

todo Finish project report

## 🎁 Why Choose Kobe?

It's FREE! 💰
Designed to be fast and easy to use.
It's your perfect companion to help manage your tasks efficiently!
Give it a try and experience the magic of stress-free task management!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

bbryant824 committed:
lvl 0

bbryant824 committed:
lv 2

bbryant824 committed:
lv 2 update 1

bbryant824 committed:
lv 1 updated

bbryant824 committed:
lv 1 updates

bbryant824 committed:
lv 1 updates

bbryant824 committed:
level-2

bbryant824 committed:
level-2 updates

bbryant824 committed:
level-3

bbryant824 committed:
level-4

bbryant824 committed:
level-5

bbryant824 committed:
level-6

bbryant824 committed:
level-6

bbryant824 committed:
level-6 updated

bbryant824 committed:
level-7 p1

bbryant824 committed:
Merge pull request #1 from bbryant824/branch-Level-7

level-7

bbryant824 committed:
level-8 p1

bbryant824 committed:
Merge pull request #2 from bbryant824/branch-Level-8

level-8 p1

bbryant824 committed:
level-8 p2, more oop

bbryant824 committed:
Merge pull request #3 from bbryant824/branch-A-MoreOOP

level-8 p2, more oop

bbryant824 committed:
level-8 p2, more oop

bbryant824 committed:
level-8 p3, A-Packages

bbryant824 committed:
Merge pull request #4 from bbryant824/add-gradle-support

Add gradle support

bbryant824 committed:
gradle updated

bbryant824 committed:
A-JUnit updated

bbryant824 committed:
A-Jar updated

bbryant824 committed:
A-JavaDoc v1

bbryant824 committed:
A-CodingStandard v1

bbryant824 committed:
Level-9, FindCommand added

bbryant824 committed:
Merge pull request #5 from bbryant824/branch-A-JavaDoc

A-JavaDoc v1

bbryant824 committed:
Merge pull request #6 from bbryant824/A-CodingStandard

A-CodingStandard v1

bbryant824 committed:
Merge pull request #7 from bbryant824/branch-Level-9

Level-9, FindCommand added

bbryant824 committed:
A-CheckStyle added

zzawook commented on src/main/java/kobe/Kobe.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class Kobe {
+
+    /** Handles storage-related operations such as saving and loading tasks. */
```
I like the detailed javadocs you provided for all variables and functions. Some hustle there!

bbryant824 replied:
yeapp thank you, will keep it up!

zzawook commented on src/main/java/kobe/Kobe.java:
```diff
@@ -0,0 +1,71 @@
+     * @param args Command line arguments (not used).
+     */
+    public static void main(String[] args) {
+        new Kobe("data/kobee.txt").run();
```
Any reason why you used plain string instead of constant/final variable?

bbryant824 replied:
i thought this was the way, didnt consider much. I will try understand and consider switching. Thanks.

zzawook commented on src/main/java/kobe/command/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+            tasks.addTask(todo);
+            ui.showTasks(tasks);
+        } else if (taskType.equals("deadline")) {
+            String[] parts = words[1].split(" /by ");
```
Should the duplicate codes / logic be extracted out as separate function?

bbryant824 replied:
yeapp, thanks for the suggestion!

zzawook reviewed (commented):
Overall, nice job on coding standard. I barely see any deviation from coding standard. Keep up the good work!

bbryant824 committed:
Merge pull request #8 from bbryant824/branch-A-CheckStyle

A-CheckStyle added

bbryant824 committed:
Level-10 v1

bbryant824 committed:
Merge pull request #9 from bbryant824/branch-Level-10

Level-10 v1

bbryant824 committed:
A-Varargs with UI and javadocs added

bbryant824 committed:
Merge pull request #10 from bbryant824/branch-A-Varargs

A-Varargs with UI and javadocs added

bbryant824 committed:
updates: modifications and packaging, fixing bugs

bbryant824 committed:
Subject: Configure Gradle build to generate JAR file with JavaFX bundled

Body:

Previously, the JAR file created did not include JavaFX dependencies, causing the application to fail on systems without JavaFX installed.

This commit configures the Gradle build to generate a fat JAR file, bundling the necessary JavaFX libraries within it. This ensures that the application can run on any system, regardless of whether JavaFX is installed.

bbryant824 committed:
Subject: Refactor TaskList#addTask to use varargs

Body:

The TaskList#addTask method previously accepted only a single task at a time, which required multiple calls to add multiple tasks.

To improve usability, we refactor the method to use Java's varargs feature, allowing multiple tasks to be added in a single method call.

bbryant824 committed:
Add assertions to Storage and AddCommand classes

The Storage and AddCommand classes lack validation checks to ensure
critical assumptions hold true during runtime.

Currently, the Storage class does not validate the presence of tasks or
check the existence of files before performing file operations.
Similarly, the AddCommand class does not validate the user input
sufficiently before adding tasks.

To improve reliability, let's add assertions to these classes:
* Ensure tasks are non-null before saving/loading
* Ensure valid task descriptions are provided in AddCommand
* Verify that task file exists before attempting to load it

These assertions will help catch potential logic errors during the
execution of the application.

bbryant824 committed:
refactor: Improve UI structure and add assertions in Main class

- Refactored `Main` class to improve code readability by modularizing layout and UI setup.
- Extracted magic numbers into constants for window dimensions and layout spacing.
- Added assertions to ensure user input is non-null and not empty before processing.
- Introduced helper method `addDialog()` to avoid code duplication when adding dialogs.
- Improved layout structure and event handler setup.

These changes enhance code quality, readability, and ensure key assumptions hold during execution.

bbryant824 committed:
refactor: Utilize Java Streams

- Refactored TaskList to use Java Streams for generating task list output.

bbryant824 committed:
Merge pull request #11 from bbryant824/branch-A-Assertion

Add assertions to Storage and AddCommand classes

bbryant824 committed:
Merge pull request #12 from bbryant824/branch-A-CodeQuality

refactor: Improve UI structure and add assertions in Main class

bbryant824 committed:
Merge pull request #13 from bbryant824/branch-A-Streams

refactor: Utilize Java Streams

bbryant824 committed:
Improved the overall user experience in the chatbot UI:

- Added the ability to close the application gracefully with a short delay after the 'bye' command.
- Centralized dialog management in addDialog() to handle both user and chatbot messages consistently.
- Applied responsive scrolling behavior to ensure the latest messages are always visible.
- Refactored the welcome message display for consistency and modularity.
- Ensured smoother transitions between commands and improved the UI layout for better user interaction.

This refactor improves the UI interaction flow and simplifies application exit behavior.

bbryant824 committed:
extension: tagging

bbryant824 committed:
update UI, README updates

bbryant824 committed:
Update README.md

bbryant824 committed:
Improve error handling and update user-friendly error messages

Currently, some error messages like "index out of bounds" do not provide
enough context for the user to understand what went wrong or how to fix
their input. Users may be confused by these technical terms.

It is important to improve the error handling to provide more
descriptive and meaningful messages. This will enhance the user
experience by guiding users to provide the correct input format.

This change:
* Adds specific error messages for each command (e.g., "mark", "delete",
  "tag") when invalid input is detected.
* Ensures that commands such as "mark" or "delete" include helpful
  instructions on what is wrong with the input and how to correct it.
* Uses `KobeException` for better distinction between input errors and
  other exceptions.
* Handles empty commands and invalid task numbers with appropriate
  feedback.

This improves usability by ensuring the user knows exactly how to fix
their mistakes rather than encountering technical error messages.

bbryant824 committed:
Enhance parser to handle more robust input validation

The parser previously did not validate input sufficiently, leading to
situations where invalid commands could cause uncaught exceptions or
incorrect feedback to users.

It is necessary to ensure that commands are validated properly before
execution to prevent unexpected crashes or misleading errors.

This update:
* Adds validation for empty inputs, task numbers, and required arguments
  for commands such as "mark", "delete", and "tag".
* Adds specific checks for "find", ensuring that users provide either a
  keyword or tag.
* Prevents the parser from accepting malformed commands by throwing
  clear, descriptive error messages via `KobeException`.

By adding these checks at the parser level, we reduce potential runtime
errors and provide more clear instructions to users when their commands
are invalid.

bbryant824 committed:
Refactor AddCommand to handle date/time parsing errors

The AddCommand class handles tasks that require date/time inputs (e.g.,
deadlines and events), but previously did not provide meaningful feedback
when date/time parsing failed.

Improving this feedback is crucial to prevent user frustration and
guiding them to input the correct format.

This change:
* Refactors `AddCommand` to catch `DateTimeParseException` and give
  users a clear message explaining the correct date/time format.
* Adds additional error handling for invalid event ranges (e.g., start
  time after end time).
* Ensures that the user receives actionable feedback if they input the
  wrong format or invalid task details.

This refactor ensures that users know exactly what went wrong and how
to fix date/time-related errors, improving usability and reducing errors.

bbryant824 committed:
Add unit tests for Todo class

The Todo class currently lacks proper unit tests, which leaves it
vulnerable to potential bugs and issues, especially as future changes
are made to the application.

Unit tests are necessary to ensure that the Todo class behaves as
expected and correctly handles various scenarios, such as creating new
todos, displaying the correct status, and ensuring task tagging works
as intended.

This update:
* Adds unit tests to check basic Todo creation and ensure that
  newly-created todos are marked as incomplete by default.
* Verifies the toString() method output to ensure it accurately reflects
  task completion and tag information.
* Adds tests to verify correct tag addition and display in Todo tasks.

By adding these tests, we improve code coverage and ensure the Todo
class performs as expected in different situations, improving code
reliability.

bbryant824 committed:
Add unit tests for Parser class

The Parser class is responsible for converting user input into commands
that can be executed. However, it currently lacks robust unit testing
to ensure that it correctly handles various valid and invalid inputs.

Adding tests for the Parser is essential to catch potential parsing
errors, especially for commands like mark, delete, and tag, which
require specific arguments.

This update:
* Adds unit tests for common commands such as `mark`, `delete`, `tag`,
  and `find` to ensure proper command parsing.
* Tests invalid input handling (e.g., empty commands, invalid task
  numbers) to verify that the Parser throws `KobeException` with
  meaningful error messages.
* Includes tests for parsing special cases such as tag commands and
  keyword searches, ensuring they behave as expected.

These tests provide better coverage for the Parser and ensure that it
functions as intended, reducing the likelihood of errors and improving
the robustness of input handling.

bbryant824 committed:
Add Javadocs for Kobe class

The Kobe class lacked comprehensive Javadocs, making it harder for
developers to understand its role and functionality in the chatbot
application. Proper documentation is necessary to maintain code clarity
and readability.

This update:
* Adds Javadocs to the Kobe class, including detailed descriptions for
  the class itself and all its methods.
* Documents the getResponse method to clarify how user input is
  processed and how commands are executed.
* Includes descriptions for the constructor and key fields (e.g., tasks,
  ui, and storage).

These Javadocs improve the overall maintainability of the code by
making it easier for future developers to understand how the Kobe class
handles task management and user interaction.

bbryant824 committed:
Add Javadocs for DialogBox class

The DialogBox class lacked proper Javadocs, making it difficult for
developers to grasp its role in displaying user and chatbot messages in
the GUI. Providing comprehensive documentation helps improve the
clarity of the UI-related code.

This update:
* Adds Javadocs for the DialogBox class, including detailed descriptions
  of its purpose and methods.
* Documents methods like getUserDialog() and getKobeDialog(), explaining
  how the dialog boxes are styled and aligned in the UI.
* Provides descriptions of how images are displayed and styled in the
  dialog boxes.

By documenting these methods and the class behavior, the DialogBox class
becomes easier to maintain and extend, especially when adding new
features or improving the UI.

bbryant824 committed:
Add Javadocs for TaskList class

The TaskList class previously lacked sufficient documentation, making it
challenging to understand the role and behavior of task manipulation
methods such as addTask, removeTask, and findTasks.

This update:
* Adds Javadocs to the TaskList class, providing clear descriptions for
  the class and its methods.
* Explains the purpose of methods like addTask(), findTasks(), and
  findTasksByTag(), ensuring that developers can easily understand how
  task lists are managed.
* Documents the structure of the task list and the methods for searching
  and modifying the list of tasks.

These improvements enhance the readability and maintainability of the
code, making it easier for developers to work with task-related
functionalities.

bbryant824 committed:
Merge branch 'master' of https://github.com/bbryant824/ip

bbryant824 committed:
Update Javadoc for AddCommand class

Added detailed Javadoc comments for the AddCommand class to explain
its purpose and functionality in adding tasks (todo, deadline, event)
to the chatbot's task list.

- Documented the constructor, explaining the fullCommand parameter and
  its role in determining the type of task to add.
- Provided thorough documentation for the execute() method, detailing
  how it handles different task types (todo, deadline, event).
- Explained the private helper methods addDeadlineTask() and
  addEventTask(), including how they parse and validate date formats
  and manage exceptions for incorrect input.
- Described error handling mechanisms in the case of invalid input and
  how exceptions are thrown with meaningful messages for the user.

This improves the clarity and maintainability of the code by making the
usage, behavior, and error handling of the class more transparent for
future contributors.

bbryant824 committed:
Update Javadoc for Kobe class

Enhanced the documentation for the Kobe class to provide a clear
understanding of its role as the main controller for the chatbot
application.

- Documented the constructor, detailing how the tasks are loaded from
  storage and initialized.
- Provided detailed explanations for the getResponse() method,
  describing how user input is processed and how different commands
  are executed.
- Included an explanation of error handling for invalid input.

By improving the Javadoc for this core class, it will be easier for
future developers to follow the flow of the application and understand
how the chatbot processes commands and interacts with the user interface.

bbryant824 committed:
Update Javadoc for ExitCommand class

Provided Javadoc comments for the ExitCommand class to explain its role
in handling user requests to terminate the chatbot.

- Documented the constructor and the execute() method, describing how
  the command sets the appropriate response in the Ui to exit the
  application.

This ensures that the ExitCommand's purpose and functionality are well
documented, making the code easier to understand and modify if needed.

bbryant824 committed:
Update Javadoc for FindCommand class

Added comprehensive Javadoc for the FindCommand class to provide clear
explanations of the class's responsibility for searching tasks by
keyword or tag.

- Documented the constructor, explaining the keyword parameter.
- Added details for the execute() method, highlighting how the
  method searches for tasks based on keyword or tag and how it handles
  no matching tasks.

Improving documentation helps ensure that future contributors
understand how the find functionality works and how it interacts
with other components.

bbryant824 committed:
Add unit tests for Parser class

The Parser class is responsible for converting user input into commands
that can be executed. However, it currently lacks robust unit testing
to ensure that it correctly handles various valid and invalid inputs.

Adding tests for the Parser is essential to catch potential parsing
errors, especially for commands like mark, delete, and tag, which
require specific arguments.

This update:
* Adds unit tests for common commands such as `mark`, `delete`, `tag`,
  and `find` to ensure proper command parsing.
* Tests invalid input handling (e.g., empty commands, invalid task
  numbers) to verify that the Parser throws `KobeException` with
  meaningful error messages.
* Includes tests for parsing special cases such as tag commands and
  keyword searches, ensuring they behave as expected.

These tests provide better coverage for the Parser and ensure that it
functions as intended, reducing the likelihood of errors and improving
the robustness of input handling.

bbryant824 committed:
Add unit tests for Todo class

The Todo class currently lacks proper unit tests, which leaves it
vulnerable to potential bugs and issues, especially as future changes
are made to the application.

Unit tests are necessary to ensure that the Todo class behaves as
expected and correctly handles various scenarios, such as creating new
todos, displaying the correct status, and ensuring task tagging works
as intended.

This update:
* Adds unit tests to check basic Todo creation and ensure that
  newly-created todos are marked as incomplete by default.
* Verifies the toString() method output to ensure it accurately reflects
  task completion and tag information.
* Adds tests to verify correct tag addition and display in Todo tasks.

By adding these tests, we improve code coverage and ensure the Todo
class performs as expected in different situations, improving code
reliability.

bbryant824 committed:
Improve error handling for mark and unmark commands

Enhanced error messages for mark and unmark commands to provide users with clear and meaningful feedback. Added validation for invalid or out-of-range task numbers, ensuring proper input formats are required. Updated Parser to handle non-numeric inputs and missing task numbers with appropriate error messages.
